src/chain.h

```diff
@@ -300,14 +308,24 @@ class CBlockIndex
         return ((nStatus & BLOCK_VALID_MASK) >= nUpTo);
     }
 
+    //! @returns true if the block is assumed-valid; this means it is queued to be
+    //!   validated by a background chainstate.
+    bool IsAssumedValid() const { return nStatus & BLOCK_ASSUMED_VALID; }
+
     //! Raise the validity level of this block index entry.
     //! Returns true if the validity was changed.
     bool RaiseValidity(enum BlockStatus nUpTo)
     {
         assert(!(nUpTo & ~BLOCK_VALID_MASK)); // Only validity flags allowed.
-        if (nStatus & BLOCK_FAILED_MASK)
-            return false;
+        if (nStatus & BLOCK_FAILED_MASK) return false;
+
         if ((nStatus & BLOCK_VALID_MASK) < nUpTo) {
+            // If this block had been marked assumed-valid and we're raising
+            // its validity to a certain point, there is no longer an assumption.
+            if (nStatus & BLOCK_ASSUMED_VALID && nUpTo >= BLOCK_VALID_SCRIPTS) {
```

Comment:
nit: an invalid nUpTo such as 6 or 7 could corrupt the result here. Consider 4|3 which is (BLOCK_VALID_CHAIN|BLOCK_VALID_TRANSACTIONS). I don't think this can actually happen though.

It might be better to explicitly check nUpTo is either exactly BLOCK_VALID_SCRIPTS?

Reply:
This is a really good point, but I think we've got bigger problems if there's an invalid `nUpTo` being passed (that's still within the `BLOCK_VALID_MASK` range as you note; for subsequent reviewers: see the assert at the beginning of the function).

I think if we were to address this, it'd be a more general fix of clamping down this function and others like `IsValid()` with a ceiling check (at the moment `BLOCK_VALID_SCRIPTS = 5`); then of course that ceiling would need to be updated in sync with any addtional `BLOCK_VALID_*` additions. So anyway I think that's best treated as a follow-up.